File: src/main/java/ru/bsearch/BSearch3.java
package ru.bsearch;

import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;

/**
 * User: dima
 * Date: Sep 24, 2010
 */
public class BSearch3 {
    private static final List<Integer> EMPTY_LIST = Collections.emptyList();

    @Test
    public void binarySearch() {
        assertThat(find(1, EMPTY_LIST), equalTo(-1));

        assertThat(find(0, Arrays.asList(1)), equalTo(-1));
        assertThat(find(1, Arrays.asList(1)), equalTo(0));
        assertThat(find(2, Arrays.asList(1)), equalTo(-1));

        assertThat(find(0, Arrays.asList(1, 2)), equalTo(-1));
        assertThat(find(1, Arrays.asList(1, 2)), equalTo(0));
        assertThat(find(2, Arrays.asList(1, 2)), equalTo(1));
        assertThat(find(3, Arrays.asList(1, 2)), equalTo(-1));

        assertThat(find(0, Arrays.asList(1, 2, 3)), equalTo(-1));
        assertThat(find(1, Arrays.asList(1, 2, 3)), equalTo(0));
        assertThat(find(2, Arrays.asList(1, 2, 3)), equalTo(1));
        assertThat(find(3, Arrays.asList(1, 2, 3)), equalTo(2));
        assertThat(find(4, Arrays.asList(1, 2, 3)), equalTo(-1));
    }

    public static int find(int value, List<Integer> values) {
        if (values.isEmpty()) return -1;

        int from = 0;
        int to = values.size();

        while (from != to) {
            int mid = (from + to) / 2;
            int midValue = values.get(mid);

            if (value == midValue) {
                return mid;
            } else if (value < midValue) {
                to = mid;
            } else if (value > midValue) {
                from = mid + 1;
            }
        }

        return -1;
    }

    public static int r_find(int value, List<Integer> values) {
        return doFind(value, 0, values.size(), values);
    }

    private static int doFind(int value, int from, int to, List<Integer> values) {
        if (from == to) return -1;

        int mid = (from + to) / 2;
        int midValue = values.get(mid);

        if (value == midValue) {
            return mid;
        } else if (value < midValue) {
            return doFind(value, from, mid, values);
        } else if (value > midValue) {
            return doFind(value, mid + 1, to, values);
        }

        throw new IllegalStateException();
    }
}

File: scala/src/ru/bsearch/BSearch3_.scala
package ru.bsearch

import org.scalatest.junit.AssertionsForJUnit
import org.junit.Test

/**
 * User: DKandalov
 */
class BSearch3_ extends AssertionsForJUnit {
  @Test def shouldFindValueInASortedList() {
    assert(find(0, List()) === -1)

    assert(find(0, List(1)) === -1)
    assert(find(1, List(1)) === 0)
    assert(find(2, List(1)) === -1)

    assert(find(0, List(1, 2)) === -1)
    assert(find(1, List(1, 2)) === 0)
    assert(find(2, List(1, 2)) === 1)
    assert(find(3, List(1, 2)) === -1)

    assert(find(0, List(1, 2, 3)) === -1)
    assert(find(1, List(1, 2, 3)) === 0)
    assert(find(2, List(1, 2, 3)) === 1)
    assert(find(3, List(1, 2, 3)) === 2)
    assert(find(4, List(1, 2, 3)) === -1)

    assert(find("1", List("0", "a", "z")) === -1)
    assert(find("0", List("0", "a", "z")) === 0)
    assert(find("a", List("0", "a", "z")) === 1)
    assert(find("z", List("0", "a", "z")) === 2)
  }

  def find[T](value: T, list: List[T], shift: Int = 0)(implicit orderer: T => Ordered[T]): Int = {
    if (list.isEmpty) return -1

    val midPos = list.size / 2
    if (value == list(midPos))
      shift + midPos
    else if (value < list(midPos))
      find(value, list.take(midPos), shift)
    else // value > midValue
      find(value, list.drop(midPos + 1), shift + midPos + 1) // didn't exclude midPos values, made quite a few attempts to understand what's wrong
  }
}